Improve active object behavior - extend active object in view direction.

lhofhansl opened the pull request:
See #6642 

**Description Updated:**
Extend the range in the players direction in which active objects stepped at the server.

Note: active objects are still active at a close range (in sphere around the player). At a distance active objects are only active a players view-cone in order to save server resource - 90-99% depending on FOV setting.

**_Was:_**
This does two things:
1. Extend the range in the players direction in which active objects stepped at the server.
2. Similar to ABM deal with too much load by skipping a few rounds. This should be generally helpful.

lhofhansl commented on src/serverenvironment.cpp:
```diff
@@ -873,10 +907,6 @@ void ServerEnvironment::activateBlock(MapBlock *block, u32 additional_dtime)
 		}
 	}
-
-	/* Handle ActiveBlockModifiers */
```
Since blocks are more frequently activated only run the ABMs at their step intervals. I think this is better anyway.

lhofhansl commented on src/serverenvironment.cpp:
```diff
@@ -1278,34 +1312,46 @@ void ServerEnvironment::step(float dtime)
 			}
 		}
+		u32 time_ms = timer.stop(true);
+		u32 max_time_ms = 5;
```
5ms tested well on my machine.
I tried with the nssm mod, which usually stops generating blocks with too many mobs. It remained playable.

sfan5 replied:
`const`

lhofhansl commented on src/serverenvironment.cpp:
```diff
@@ -279,30 +279,67 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
+		fillRadiusBlock(pos, active_block_range, m_abm_list);
+		fillRadiusBlock(pos, active_block_range, newlist);
+
+		s16 player_ao_range = std::min(active_object_range, playersao->getWantedRange());
```
Note that with default settings this is not called.

lhofhansl commented on src/serverenvironment.cpp:
```diff
@@ -279,30 +279,67 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
+	for(const PlayerSAO *playersao : active_players) {
+		v3s16 pos = getNodeBlockPos(floatToInt(playersao->getBasePosition(), BS));
+		fillRadiusBlock(pos, active_block_range, m_abm_list);
+		fillRadiusBlock(pos, active_block_range, newlist);
```
Yes, this will keep vectors twice, but this made the rest of the changes smaller (with default setting we're talking 216 vectors). Could of course refactor ActiveBlockList and avoid the extra references. Not sure it would actually be more efficient.

sfan5 commented on src/serverenvironment.h:
```diff
@@ -422,6 +424,8 @@ class ServerEnvironment : public Environment
 	IntervalLimiter m_active_blocks_nodemetadata_interval;
 	int m_active_block_interval_overload_skip = 0;
+	int m_active_object_interval_overload_skip = 0;
+	float m_accdtime = 0;
```
what time? add a comment please

lhofhansl commented:
Actually now I think the separation of ABMs and active_objects is not good. mobs_redo for example registers ABMs in order for mobs to spawn. Now the further distance is used to see existing mobs, but no new mobs would spawn in the wider area.

SmallJoker commented:
This feels like quantum mechanics, where you have to look at it to make it work as expected.
What I want to say with that: Objects will only be active as long you watch them, so there's a higher chance for you to get attacked in front than on the back, because the entities will be enabled there.
For some objects it's wanted that they continue working without looking at them (pipeworks, for example), which is changed by this PR, especially when zooming.
Ignoring the FOV impact, separating the ABM and active object list to handle different block ranges is a very good idea, so spawned entities/mobs around the players will stay active in that range, no matter what the player does.

paramat commented:
I think i agree with SmallJoker.

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -279,30 +279,67 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
-			}
+	for(p.Y=p0.Y-r; p.Y<=p0.Y+r; p.Y++)
+	for(p.Z=p0.Z-r; p.Z<=p0.Z+r; p.Z++)
+	{
```
wrong code style

lhofhansl replied:
This is explicitly allow, no?

SmallJoker replied:
The nested loop is fine. Wrong formatted lines (spaces, indents or pointers) should be corrected if they're touched in a change.

paramat replied:
Spaces after 'for', spaces around '=', '<=', '+' and '-', and opening brace on previous line.

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -279,30 +279,67 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
+	for(p.Y=p0.Y-r; p.Y<=p0.Y+r; p.Y++)
+	for(p.Z=p0.Z-r; p.Z<=p0.Z+r; p.Z++)
+	{
+		if (isBlockInSight(p, camera_pos, camera_dir, camera_fov, r * BS * MAP_BLOCKSIZE)) {
```
unroll r * BS * MAP_BLOCKSIZE it's not modified and will be a little bit faster too :)

lhofhansl replied:
I had a version that did this, then I assumed the compiler would deduce this anyway. (r is const)
Happy to change it.

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -279,30 +279,67 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
-	for (const v3s16 &active_position : active_positions) {
-		fillRadiusBlock(active_position, radius, newlist);
+	m_abm_list = m_forceloaded_list;
+	for(const PlayerSAO *playersao : active_players) {
```
missing space

lhofhansl commented:
@SmallJoker @paramat , they would only be de-activated at a distance. Close by they are still subject to the active block range.

There are two ranges: 
1. The active block range. Here ABMs _and_ active objects are active. This is always a sphere around each player.
2. The extended range for active objects. There only active objects are active and only where the player looks.

So no, they do not all go away when you don't look, and when you're close they be active whether you look or not.

That seems to be a good compromise between function and performance. Within a defined radius (all around) active objects are always active. Past that distance they are active further out (assuming they cannot attack from 6 map blocks away) in the view cone.

Note with standard FOV we save 90% of the extra server work or 99.6% at zoom FOV 15.

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -279,30 +279,67 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
+	for(p.X=p0.X-r; p.X<=p0.X+r; p.X++)
+	for(p.Y=p0.Y-r; p.Y<=p0.Y+r; p.Y++)
+	for(p.Z=p0.Z-r; p.Z<=p0.Z+r; p.Z++)
+	{
```
Code style needs correction here too.
`for (p.Z = p0.Z - r; p.Z <= p0.Z + r; p.Z++) {`

paramat replied:
As my comment above. Brace on previous line.

paramat commented:
>  they would only be de-activated at a distance. Close by they are still subject to the active block range.

Yes i know, sorry my comment may have given the wrong impression. I am still unsure about view direction affecting active objects, but don't feel strongly about it.

lhofhansl commented:
I also found that the active object skipping I added might be too aggressive.

Played around with TNT (thinking that'd be about the worst case), and indeed with a larger blast radius the active object can be busy for 1s or more (that's bad, but that's a different story). With the 5ms time check this would now skip the active object for 200 rounds, or about 20s.

Need to either remove it, or be smarter. Perhaps keep the moving average of the execution time and use that. That same applies to the existing ABM code btw. only the intervals are less critical and longer there.

lhofhansl commented:
OK... This version addresses review comments. It also improves the time budget:
1. Add an limiter to active objects
2. When ABMs and AOs are over budgets, increase the limiter budget (instead of skipping a number of steps). That makes it more smooth depending on how far over budget the process it.
3. Do not hardcode the budget limit. Take 1/5 of the allowed budget (the 1/5 is still hardcoded, though)

paramat commented:
```
#    From how far clients know about objects, stated in mapblocks (16 nodes).
#    type: int
# active_object_send_range_blocks = 3

#    How large area of blocks are subject to the active block stuff, stated in mapblocks (16 nodes).
#    In active blocks objects are loaded and ABMs run.
#    type: int
# active_block_range = 3
```
Before a merge this documentation should be updated to explain the new behaviour.
Both set the same (the current default) doesn't change behaviour. I don't have any objection to the PR, just thinking about possible implications of view-determined object activation =)

lhofhansl commented:
Done.
Is there any point in splitting this into multiple PRs. One for the time budgets changes and one of the extended view range...

paramat commented:
That would be ideal but i personally won't insist on it.

lhofhansl commented:
Let's do it one. Easier that way.
Anything I need to do?

lhofhansl commented:
Rebased.
Also using a moving average now to calculate the time taken in stepping active object, since that can be very bursty.

Should be all good now.

paramat commented:
BTW #6683 but if you are busy maybe i can fix those.

lhofhansl commented:
Looked into #6683 , it's actually not a bug (or caused by my changes). It's working as designed - perhaps a bit unexpected for very small FOVs.

paramat commented:
I concept-approve this PR but am not qualified to do a complete review, some of this is beyond me.

lhofhansl commented:
I split this form #6721 .

#6721 , #6723 and this one work best together.

paramat commented:
I approve the concept (Not an official approval) of:

> Extend the range in the players direction in which active objects stepped at the server.

Don't feel qualified to judge the skipping of rounds.

lhofhansl commented:
@paramat That part is in #6721 now.

For games with complex mobs (like NSSM) it makes the difference between playable and not playable. In 90% of the cases there's no difference anyway.

I've have all these three on, and (apart from what @Fixer-007 notices, and I have since fixed) and seen no issues, other than the jitter goes away and there's more server time for generating the world.

paramat commented:
Aha ok, i'll try reviewing this one.

paramat commented on src/serverenvironment.cpp:
```diff
@@ -279,30 +279,66 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
+	const s16 r_nodes = r * BS * MAP_BLOCKSIZE;
+	for (p.X = p0.X - r; p.X <= p0.X+r; p.X++)
+	for (p.Y = p0.Y - r; p.Y <= p0.Y+r; p.Y++)
+	for (p.Z = p0.Z - r; p.Z <= p0.Z+r; p.Z++) {
```
The 3 lines above: spaces around '+'

paramat commented on src/serverenvironment.cpp:
```diff
@@ -1149,18 +1178,21 @@ void ServerEnvironment::step(float dtime)
 			Update list of active blocks, collecting changes
 		*/
+		// use active_object_send_range_blocks since that is max distance
+		// for active objects sent the client anyway
```
'sent to'

paramat commented:
I'm :+1:  with everything up to the 'overload skip' stuff, which is where i'm struggling to completely understand. So i'd like a third dev to look at this.

lhofhansl commented:
Lemme update the patch here. I split off stuff from this PR, but neglected to update this one :(

lhofhansl commented:
@paramat Updated now.

paramat commented:
:+1:

lhofhansl commented:
Great. Thanks. Will merge in a few.

lhofhansl commented:
5a03b1f5
